cocotb/types/array.py

```diff
@@ -0,0 +1,233 @@
+            stop = item.stop or self.right
+            if item.step is not None:
+                raise IndexError("do not specify step")
+            start_i = self._translate_index(start)
+            stop_i = self._translate_index(stop)
+            if start_i > stop_i:
+                raise IndexError("slice direction does not match array direction")
+            # avoid second copy
+            cls = type(self)
+            val = cls.__new__(cls)
+            val._value = self._value[start_i : stop_i + 1]
+            val._range = Range(start, self.direction, stop)
+            return val
+        raise TypeError(
+            "indexes must be ints or slices, not {}".format(type(item).__name__)
+        )
+
+    @overload
+    def __setitem__(self, item: int, value: Any) -> None:
+        pass  # pragma: no cover
+
+    @overload
+    def __setitem__(self, item: slice, value: Iterable[Any]) -> None:
+        pass  # pragma: no cover
+
+    def __setitem__(self, item, value):
+        if isinstance(item, int):
+            idx = self._translate_index(item)
+            self._value[idx] = value
+        elif isinstance(item, slice):
+            start = item.start or self.left
+            stop = item.stop or self.right
+            if item.step is not None:
+                raise IndexError("do not specify step")
+            start_i = self._translate_index(start)
+            stop_i = self._translate_index(stop)
+            if start_i > stop_i:
+                raise IndexError("slice direction does not match array direction")
+            value = list(value)
+            if len(value) != (stop_i - start_i + 1):
+                raise ValueError("value is not the same length as the slice")
+            self._value[start_i : stop_i + 1] = value
+        else:
+            raise TypeError(
+                "indexes must be ints or slices, not {}".format(
+                    type(item).__name__
+                )
+            )
+
+    def __repr__(self) -> str:
+        return "{}({!r}, {!r})".format(type(self).__name__, self._value, self._range)
+
+    def index(self, value: Any, start: Optional[int] = None, stop: Optional[int] = None) -> int:
+        """Return index of first occurence of value."""
+        if start is not None:
+            start = self._translate_index(start)
+        else:
+            start = 0
+        if stop is not None:
+            stop = self._translate_index(stop)
+        else:
+            stop = maxsize  # same default value used by Python lists
+        idx = self._value.index(value, start, stop)
+        return self._range[idx]
+
+    def count(self, value: Any) -> int:
+        """Return number of occurences of value."""
+        return self._value.count(value)
+
+    def _translate_index(self, item: int) -> int:
+        try:
+            return self._range.index(item)
+        except ValueError:
+            raise IndexError("index {} out of range".format(item)) from None
```

Comment:
```suggestion
            raise IndexError(f"index {item} out of range") from None
```

Reply:
Do you have a particular opinion on when to use f strings vs format?

Reply:
I'd do it like pyupgrade does it, only for "simple expressions" - actually, it would be easiest to just run it on the code.